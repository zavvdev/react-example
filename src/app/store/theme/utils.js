import { localStorageService } from "app/services/LocalStorageService";
import { IS_DARK_THEME_LOCAL_STORAGE_KEY } from "app/store/theme/config";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(IS_DARK_THEME_LOCAL_STORAGE_KEY, `${darkModeState}`);
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    IS_DARK_THEME_LOCAL_STORAGE_KEY,
  );
  return !!JSON.parse(rawStorageValue);
};
