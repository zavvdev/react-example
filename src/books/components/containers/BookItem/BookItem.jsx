import { Button, Typography } from "books/gateway/input";
import PropTypes from "prop-types";
import { useBookItemStyles } from "books/components/containers/BookItem/BookItem.styles";
import { useTranslation } from "react-i18next";
import { BOOKS_I18N_NAMESPACE } from "books/i18n";

export function BookItem({
  title,
  author,
  date,
  price,
  cover,
  onAddToCart,
  onRemoveFromCart,
  isInCart,
}) {
  const { t } = useTranslation(BOOKS_I18N_NAMESPACE);
  const classes = useBookItemStyles();

  return (
    <div className={classes.root}>
      <img src={cover} alt="title" className={classes.cover} />
      <div>
        <Typography className={classes.title} tag="h2">
          {title}
        </Typography>
        <div className={classes.infoWrap}>
          <Typography tag="i">{date}, </Typography>
          <Typography tag="span">
            {t("bookItem.author", {
              author,
            })}
          </Typography>
        </div>
        <Typography className={classes.price} tag="div">
          {price}
        </Typography>
        <Button
          className={classes.addToCartBtn}
          onClick={isInCart ? onRemoveFromCart : onAddToCart}
        >
          {isInCart ? t("bookItem.alreadyInCart") : t("bookItem.addToCart")}
        </Button>
      </div>
    </div>
  );
}

BookItem.propTypes = {
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  cover: PropTypes.string.isRequired,
  onAddToCart: PropTypes.func.isRequired,
  onRemoveFromCart: PropTypes.func.isRequired,
  isInCart: PropTypes.bool.isRequired,
};
