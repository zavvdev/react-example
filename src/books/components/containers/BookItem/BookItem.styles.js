import { createUseStyles } from "react-jss";

export const useBookItemStyles = createUseStyles((theme) => ({
  root: {
    display: "flex",
    marginBottom: "26px",
    paddingBottom: "16px",
    borderBottom: `1px solid ${theme.borderColor}`,
    "@media screen and (max-width: 500px)": {
      display: "block",
    },
  },
  cover: {
    height: "210px",
    marginRight: "16px",
    marginBottom: "10px",
  },
  title: {
    margin: "0 0 10px 0",
    fontSize: "20px",
    lineHeight: "25px",
  },
  infoWrap: {
    marginBottom: "16px",
  },
  price: {
    fontSize: "30px",
    fontWeight: "bolder",
  },
  addToCartBtn: {
    marginTop: "16px",
  },
}));
