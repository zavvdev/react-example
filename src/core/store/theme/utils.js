import { LOCAL_STORAGE_KEYS } from "core/config/localStorage";
import { localStorageService } from "core/services/LocalStorageService";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
    `${darkModeState}`,
  );
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
  );
  return !!JSON.parse(rawStorageValue);
};
