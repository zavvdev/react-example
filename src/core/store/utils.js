import { all, call, spawn } from "redux-saga/effects";
import { errorTrackingService } from "core/services/ErrorTrackingService";

export const combineMiddleware = (middlewareList) => function* combiner() {
  const getRootGenerator = (middleware) => function* rootGenerator() {
    while (true) {
      try {
        yield call(middleware);
        break;
      } catch (e) {
        errorTrackingService.reportError(e);
      }
    }
  };

  yield all(middlewareList.map((middleware) => {
    return spawn(getRootGenerator(middleware));
  }));
};
