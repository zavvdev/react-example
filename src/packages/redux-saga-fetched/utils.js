import { ACTION_TYPES, DATA_STATUS_TYPES } from "packages/redux-saga-fetched/config";

export const createActionType = ({ createdKey, actionTypePattern }) => {
  return `${createdKey}_${actionTypePattern}`;
};

export const createKey = (key) => key.join("_");

export const createActionTypePatterns = (domain) => ({
  request: `${domain}@${ACTION_TYPES.request}`,
  success: `${domain}@${ACTION_TYPES.success}`,
  failure: `${domain}@${ACTION_TYPES.failure}`,
  invalidate: `${domain}@${ACTION_TYPES.invalidate}`,
  reset: `${domain}@${ACTION_TYPES.reset}`,
});

export const createState = ({
  isLoading,
  isFetching,
  isLoaded,
  isError,
  isValid,
  status,
  data,
}) => ({
  isLoading,
  isFetching,
  isLoaded,
  isError,
  isValid,
  status,
  data,
});

export const createResetState = () => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: false,
  isError: false,
  isValid: false,
  status: DATA_STATUS_TYPES.reset,
  data: null,
});

export const createRequestState = ({ state, payload }) => {
  const isLoading = !state[payload.createdKey]?.data;
  const isFetching = !!state[payload.createdKey]?.data;
  return createState({
    isLoading,
    isFetching,
    isLoaded: false,
    isError: false,
    isValid: false,
    status: isLoading ? DATA_STATUS_TYPES.loading : DATA_STATUS_TYPES.fetching,
    data: state[payload.createdKey]?.data || null,
  });
};

export const createSuccessState = ({ payload }) => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: true,
  isError: false,
  isValid: true,
  status: DATA_STATUS_TYPES.loaded,
  data: payload.data,
});

export const createFailureState = ({ state, payload }) => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: false,
  isError: true,
  isValid: false,
  status: DATA_STATUS_TYPES.error,
  data: state[payload.createdKey]?.data || null,
});

export const createInvalidateState = ({ state, payload }) => createState({
  isLoading: state[payload.createdKey]?.isLoading,
  isFetching: state[payload.createdKey]?.isFetching,
  isLoaded: state[payload.createdKey]?.isLoaded,
  isError: state[payload.createdKey]?.isError,
  isValid: false,
  status: state[payload.createdKey]?.status,
  data: state[payload.createdKey]?.data,
});
