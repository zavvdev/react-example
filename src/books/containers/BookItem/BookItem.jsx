import PropTypes from "prop-types";
import { Shared } from "app/shared";
import { useBookItemStyles } from "books/containers/BookItem/BookItem.styles";
import { useBooksTranslation } from "books/i18n/useBooksTranslation";

export function BookItem({
  title,
  author,
  date,
  price,
  cover,
  onAddToCart,
  onRemoveFromCart,
  isInCart,
}) {
  const { t } = useBooksTranslation();
  const classes = useBookItemStyles();

  return (
    <div className={classes.root}>
      <img src={cover} alt={title} className={classes.cover} />
      <div>
        <h2 className={classes.title}>{title}</h2>
        <div className={classes.infoWrap}>
          <i>{date}, </i>
          <span>
            {t("bookItem.author", {
              author,
            })}
          </span>
        </div>
        <div className={classes.price}>{price}</div>
        <Shared.Button
          className={classes.addToCartBtn}
          onClick={isInCart ? onRemoveFromCart : onAddToCart}
        >
          {isInCart ? t("bookItem.alreadyInCart") : t("bookItem.addToCart")}
        </Shared.Button>
      </div>
    </div>
  );
}

BookItem.propTypes = {
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  cover: PropTypes.string.isRequired,
  onAddToCart: PropTypes.func.isRequired,
  onRemoveFromCart: PropTypes.func.isRequired,
  isInCart: PropTypes.bool.isRequired,
};
