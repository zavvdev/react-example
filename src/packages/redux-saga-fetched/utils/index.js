import { ACTION_TYPES, EFFECT_TYPES } from "packages/redux-saga-fetched/config";

export const createActionType = ({ createdKey, actionTypePattern }) => {
  return `${createdKey}_${actionTypePattern}`;
};

export const createKey = (key) => key.join("_");

export const createActionTypePatterns = (domain) => ({
  [EFFECT_TYPES.query]: {
    request: `${domain}@${ACTION_TYPES[EFFECT_TYPES.query].request}`,
    success: `${domain}@${ACTION_TYPES[EFFECT_TYPES.query].success}`,
    failure: `${domain}@${ACTION_TYPES[EFFECT_TYPES.query].failure}`,
    invalidate: `${domain}@${ACTION_TYPES[EFFECT_TYPES.query].invalidate}`,
    reset: `${domain}@${ACTION_TYPES[EFFECT_TYPES.query].reset}`,
  },
  [EFFECT_TYPES.mutation]: {
    request: `${domain}@${ACTION_TYPES[EFFECT_TYPES.mutation].request}`,
    success: `${domain}@${ACTION_TYPES[EFFECT_TYPES.mutation].success}`,
    failure: `${domain}@${ACTION_TYPES[EFFECT_TYPES.mutation].failure}`,
    reset: `${domain}@${ACTION_TYPES[EFFECT_TYPES.mutation].reset}`,
  },
});
