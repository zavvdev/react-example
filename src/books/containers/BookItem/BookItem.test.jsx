import { fireEvent, screen } from "@testing-library/react";
import { BookItem } from "books/containers/BookItem/BookItem";
import { testUtils } from "__tests__/utils";

describe("BookItem container", () => {
  test("should render BookItem which has isInCart=false", () => {
    const result = testUtils.render(
      <BookItem
        title="Mock"
        author="mock"
        date="12/12/22"
        price="$42"
        cover="https://mock.com/42"
        onAddToCart={() => {}}
        onRemoveFromCart={() => {}}
        isInCart={false}
      />,
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("should render BookItem which has isInCart=true", () => {
    const result = testUtils.render(
      <BookItem
        title="Mock"
        author="mock"
        date="12/12/22"
        price="$42"
        cover="https://mock.com/42"
        onAddToCart={() => {}}
        onRemoveFromCart={() => {}}
        isInCart
      />,
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("should call onAddToCart function", () => {
    const onAddToCartMock = jest.fn();
    testUtils.render(
      <BookItem
        title="Mock"
        author="mock"
        date="12/12/22"
        price="$42"
        cover="https://mock.com/42"
        onAddToCart={onAddToCartMock}
        onRemoveFromCart={() => {}}
        isInCart={false}
      />,
    );
    const addToCartButton = screen.getByText("bookItem.addToCart");
    fireEvent.click(addToCartButton);
    expect(onAddToCartMock).toBeCalled();
  });

  test("should call onRemoveFromCart function", () => {
    const onRemoveFromCartMock = jest.fn();
    testUtils.render(
      <BookItem
        title="Mock"
        author="mock"
        date="12/12/22"
        price="$42"
        cover="https://mock.com/42"
        onAddToCart={() => {}}
        onRemoveFromCart={onRemoveFromCartMock}
        isInCart
      />,
    );
    const removeFromCartButton = screen.getByText("bookItem.alreadyInCart");
    fireEvent.click(removeFromCartButton);
    expect(onRemoveFromCartMock).toBeCalled();
  });
});
