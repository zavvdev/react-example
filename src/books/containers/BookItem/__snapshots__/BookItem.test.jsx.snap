// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BookItem which has isInCart=false 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <img
      alt="Mock"
      class="cover"
      src="https://mock.com/42"
    />
    <div>
      <h2
        class="title"
      >
        Mock
      </h2>
      <div
        class="infoWrap"
      >
        <i>
          12/12/22, 
        </i>
        <span>
          bookItem.author
        </span>
      </div>
      <div
        class="price"
      >
        $42
      </div>
      <button
        class="root addToCartBtn"
        type="button"
      >
        bookItem.addToCart
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render BookItem which has isInCart=true 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <img
      alt="Mock"
      class="cover"
      src="https://mock.com/42"
    />
    <div>
      <h2
        class="title"
      >
        Mock
      </h2>
      <div
        class="infoWrap"
      >
        <i>
          12/12/22, 
        </i>
        <span>
          bookItem.author
        </span>
      </div>
      <div
        class="price"
      >
        $42
      </div>
      <button
        class="root addToCartBtn"
        type="button"
      >
        bookItem.alreadyInCart
      </button>
    </div>
  </div>
</DocumentFragment>
`;
